Create stocktake from master list

mark-prins opened the pull request:
Fixes half of #938 

Only 14 files, but enough for now. Will add in the locations part in a separate PR.

Changes in this PR: 
- Removes the **Select Item** modal which was shown when creating a stocktake
- Gives the option to select a master list
- Creates the stocktake and adds lines for 
  - each item on the master list which has no stock
  - each stock line for each item on the list
- Adds a message to say that you cannot edit finalised stocktakes ( perhaps should be applied to other areas too 🤔 )
- removes the **Locked** control for finalised stocktakes
- Enables a disabled state on the dropdown control ( bonus feature! )

Here's how it looks:
 
<img width="744" alt="Screenshot 2023-02-14 at 2 16 02 PM" src="https://user-images.githubusercontent.com/9192912/218622650-f418aad6-fdb2-4b38-bffb-ec03e6d6a429.png">

<img width="1175" alt="Screenshot 2023-02-14 at 3 13 39 PM" src="https://user-images.githubusercontent.com/9192912/218622633-11297769-7cb6-473e-bcd4-7aec5c6e9067.png">

### Testing

There are some unit tests to check the creation logic. 
To test manually, create a stocktake.
- [ ] Without selecting a master list : should create an empty stocktake
- [ ] With selecting a master list : should create stocktake rows 
- [ ] Check that rows are created for each stock line
- [ ] and for items with no stock
- Open a finalised stocktake and check that
  - [ ] the **Locked** control is gone
  - [ ] there is an info message showing

CarlosNZ commented:
When I set the "Locked" flag, the stocktake gets finalised and I can no longer unlock it. Don't think this was intentional?

mark-prins commented:
> When I set the "Locked" flag, the stocktake gets finalised and I can no longer unlock it. Don't think this was intentional?

Good catch thanks! have added back the button if the stocktake is on hold, and added an info panel message for the 'on hold' status

andreievg commented on server/service/src/stocktake/insert.rs:
```diff
@@ -110,6 +119,90 @@ pub fn insert_stocktake(
+        sell_price_per_pack: None,
+    };
+    if let Some(stock_line) = stock_line {
+        let stock_line_row = stock_line.stock_line_row.clone();
```
I like destructuring to make sure things are not missed, like `location_id` (initially and in the future when/if new fields are added), even though it's more vertical code, it's generally better imo, also I think this is a use case for splitting the method into two. As an example, together with previous comments (just cause it seemed fun to do):

https://github.com/openmsupply/open-msupply/compare/938-create-stocktake...938-example

andreievg commented on server/service/src/stocktake/insert.rs:
```diff
@@ -110,6 +119,90 @@ pub fn insert_stocktake(
+            .unwrap();
+
+        if stock_lines.len() == 0 {
+            insert_stocktake_line(&connection, stocktake_id, &item_id, None);
```
do we need to insert empty line ? we don't seem to do this in mSupply ¯\_(ツ)_/¯

mark-prins replied:
it was a specific request - we didn't originally insert empty lines and an issue was raised for them to be added, around the time of Amb Vic testing

mark-prins replied:
The reason is that if you only create stocktake lines for items with stock_lines, you don't have all items on the stocktake.

e.g. I used this master list

<img width="707" alt="Screenshot 2023-02-17 at 11 13 03 AM" src="https://user-images.githubusercontent.com/9192912/219499015-a684bb25-5821-44fc-8fdd-062fafc45801.png">

to create this stocktake

<img width="614" alt="Screenshot 2023-02-17 at 11 13 25 AM" src="https://user-images.githubusercontent.com/9192912/219499032-0a31d529-e823-4e4a-9bc9-82c5e66b5c26.png">

andreievg reviewed (changes requested):
Had a quick look, change request is mainly from moving logic to generate method, but also had some questions in there.

Also do we need to validate that master list belong to store_id ? (I would assume this would already be filtered by store in front end, but as a double check I think we should, we did for other things, especially if we end up adding empty `placeholder` if stocklines don't exists)

mark-prins commented:
Thanks @andreievg I've implemented those changes, and added a test for a valid masterlist. 
The unit tests fail on the example branch, so I'll have to check that before going further.

mark-prins commented:
🙏  Thanks @andreievg - I've refactored to use destructuring and implement a generate fn

I have made your example less elegant by adding back the non-stock rows, otherwise items aren't added to the stocktake 😱 

And in the refactor, picked up that the stock_lines were not being restricted to the current store - the stocktake was being populated with stock line details from other stores if the item was visible in the current store. Fixed that, and tested. Will add in a unit test now.

andreievg reviewed (approved):
🙏 for changes, looks great, had 2 questions but nothing stopping merge

mark-prins commented on client/packages/inventory/src/Stocktake/ListView/CreateStocktakeButton.tsx:
```diff
@@ -1,41 +1,150 @@
+    onClose,
+    disableBackdrop: true,
+  });
+  const isLoading = isLoadingMasterLists;
```
yes, good question! in isolation this doesn't make any sense.. it's basically setting up for the 'create by location' part as well, where the `isLoading` is a combination of either `isLoadingMasterLists || isLoadingLocations`

mark-prins commented on client/packages/inventory/src/Stocktake/ListView/CreateStocktakeButton.tsx:
```diff
@@ -1,41 +1,150 @@
+}
+
+const DEFAULT_ARGS: CreateStocktakeArgs = {
+  masterListId: 'undefined',
```
will have a re-look at the select. I started with `undefined` but mui complains if you switch from uncontrolled ( which `undefined` will give you ) to controlled ( a string value, i.e. `'undefined'` ) more importantly though - if the value is `undefined` then it will not show the selected option, which in this case is the 'please select' prompt.

I wonder about pushing that logic down to the `Select` itself, and removing the weirdness from the `CreateStocktakeButton` component. will have a go..
